### PR TITLE
[billing] Create Stripe invoices for teams based on their usage

### PR DESCRIPTION
**Pull request by andrew-farries**
## Description

Update teams' Stripe subscriptions with their usage credits during a given billing period.

Each time the usage reconciler runs, update Stripe invoices with usage credits for each team.

An example invoice: 

image

(exact dollar amounts per credit are still TBD, the value is a placeholder).

## Related Issue(s)
Part of #9036 

## How to test
1. Open a preview environment for this PR.
2. Create a team called 'Gitpod n', where `n` is the lowest number not yet used.
3. Enable usage based billing for the new team.
4. Start a workspace.
5. Update the `usage` component to make it run more frequently:
  * `kubectl edit deploy/usage`
  * Edit the `usage` container's `args` field to look like:
    ```yaml
     - args:
       - run
       - --schedule
       - 1m
    ```
6. Wait until the usage reconciler runs (tail its logs).
6. Visit the Stripe dashboard and find the Stripe customer that corresponds to the team you created.
7. Click on the Customer's subscription.
8. Check the 'Upcoming invoice' section and see that the customer has used Gitpod credits.

Every time the reconciler runs, the invoice should be updated with another credit (assuming the above interval of `1m` for the reconciler).

## Release Notes

```release-note
NONE
```

* [x] /werft with-payment=true

### CODE REVIEW

**Comment by jankeromnes** [comment]
Epic! 🔥 I can take a look. 👀

File: components/usage/pkg/stripe/stripe.go [hunk 1]
```diff
@@ -36,19 +38,54 @@ func Authenticate(apiKeyFile string) error {
 	return nil
 }
 
-// FindCustomersForTeamIds queries the stripe API to find all customers with a teamId in `teamIds`.
-func FindCustomersForTeamIds(teamIds []string) {
+// UpdateUsage updates teams' Stripe subscriptions with usage data
+// `usageForTeam` is a map from team name to total workspace seconds used within a billing period.
```
Comment (jankeromnes) [inline components/usage/pkg/stripe/stripe.go:42]: I think `usageForTeam` will soon become a bit more complex, due to there being multiple workspace sizes/classes, which would consume credits at different rates.

But for now it's fine to have them as just seconds without workspace class information.

File: components/usage/pkg/stripe/stripe.go [hunk 2]
```diff
@@ -72,3 +109,8 @@ func queriesForCustomersWithTeamIds(teamIds []string) []string {
 
 	return queries
 }
+
+// workspaceSecondsToCredits converts seconds (of workspace usage) into Stripe credits.
+func workspaceSecondsToCredits(seconds int64) int64 {
+	return (seconds + 59) / 60
```
Comment (jankeromnes) [inline components/usage/pkg/stripe/stripe.go:115]: Oh, that's an interesting way to round up. I guess this is equivalent to something like `Math.ceil(seconds / 60)`?

**Review by jankeromnes — APPROVED** [review]
Change looks good and works like a charm! ✨ So cool to see this all coming together.

Added a few optional comments/thoughts/questions, but please feel free to dismiss any of them and merge as is. 🚢

/hold

File: components/usage/pkg/controller/reconciler.go [hunk 3]
```diff
@@ -132,18 +132,13 @@ func (u *UsageReconciler) ReconcileTimeRange(ctx context.Context, from, to time.
 }
 
 func submitUsageReport(report []TeamUsage) {
-	var teamIdSet = make(map[string]bool)
-
-	// Convert the usage report into a set of teamIds occurring in the report
+	// Convert the usage report to sum all entries for the same team.
+	var summedReport = make(map[string]int64)
```
[the remark this answers is not in the document]
Comment (andrew-farries) [reply components/usage/pkg/controller/reconciler.go:136]: We're using `int64` elsewhere to store workspace runtimes, which already gives us *way* more range on the positive side than we need to store monthly usage, so I don't think a switch to `uint64` is necessary.

File: components/usage/pkg/stripe/stripe.go [hunk 4]
```diff
@@ -36,19 +38,54 @@ func Authenticate(apiKeyFile string) error {
 	return nil
 }
 
-// FindCustomersForTeamIds queries the stripe API to find all customers with a teamId in `teamIds`.
-func FindCustomersForTeamIds(teamIds []string) {
+// UpdateUsage updates teams' Stripe subscriptions with usage data
+// `usageForTeam` is a map from team name to total workspace seconds used within a billing period.
+func UpdateUsage(usageForTeam map[string]int64) error {
```
[the remark this answers is not in the document]
Comment (andrew-farries) [reply components/usage/pkg/stripe/stripe.go:43]: See [this comment](https://github.com/gitpod-io/gitpod/pull/10713#discussion_r899934544)

File: components/usage/pkg/stripe/stripe.go (hunk 2 above)
Comment (andrew-farries) [reply components/usage/pkg/stripe/stripe.go:115]: yes, this is an easy way to ensure that we bill partial minutes as full minutes.

File: components/usage/pkg/controller/reconciler.go [hunk 5]
```diff
@@ -132,18 +132,13 @@ func (u *UsageReconciler) ReconcileTimeRange(ctx context.Context, from, to time.
...
 	for _, usageEntry := range report {
-		teamIdSet[usageEntry.TeamID] = true
-	}
-	teamIds := make([]string, 0, len(teamIdSet))
-	for k := range teamIdSet {
-		teamIds = append(teamIds, k)
+		summedReport[usageEntry.TeamID] += usageEntry.WorkspaceSeconds
```
[the remark this answers is not in the document]
Comment (andrew-farries) [reply components/usage/pkg/controller/reconciler.go:138]: yes, maps are initialised such that values have the default value for their type.

**Pull request merged**